File: apiome-rest/src/app/export_roundtrip.py
"""On-demand export round-trip comparison — IXH-4.4 (#5112).

The strongest possible answer to *"is this export honest?"* is empirical: emit the
artifact, re-import it through the matching import adapter, and diff the re-imported
canonical model against the source revision. The IXH-1.7 conformance matrix
(:mod:`app.roundtrip_matrix`) proves this claim in CI over the corpus; this module
runs the **same loop for one user document, on demand**, so the Export Studio can
show the user their own round-trip evidence instead of asking them to trust a grid
they cannot see their document in.

The composition deliberately reuses the 1.7 seams rather than re-deriving them, so a
Studio result reconciles with the published matrix for corpus entries:

1. **emit + predict** — :func:`app.export_dispatch.dispatch_from_source` produces the
   emitted bundle *and* the fidelity report from one snapshot of the source, so the
   round-trip verdict and the on-screen fidelity/heatmap describe the same emit
   (unlike :func:`app.roundtrip_matrix.run_roundtrip`, which recomputes its own
   report for the CI matrix);
2. **gate** — :func:`app.roundtrip_matrix.import_adapter_for_emit` /
   :func:`~app.roundtrip_matrix.classify_cell` decide whether the loop can close at
   all; a target with no import adapter is **skipped with the matrix's own
   explanation**, never silently omitted;
3. **re-import** — :func:`app.roundtrip_matrix.reimport_emitted` parses + normalizes
   the emitted bundle back through the matching adapter;
4. **diff + reconcile** — :func:`app.import_source.canonical_diff` compares the two
   models and :func:`app.roundtrip_matrix.reconcile` joins every difference to the
   fidelity finding that explains it. A difference the report explains is
   **expected**; an unexplained difference (or an ``OK`` over-claim) is a fidelity
   bug the user should report.

The run is explicit and bounded: it happens only when a caller posts to the route,
performs exactly one emit + one re-import, and persists nothing (read-only emit, no
job row, no field-identity rows).
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field

from . import __version__
from .capability_registry import REGISTRY_VERSION
from .emitter import EmitterTarget, describe_emit_targets, get_emitter
from .export_dispatch import dispatch_from_source
from .export_source import ExportSource
from .import_source import (
    CanonicalDiffEntry,
    ImportSourceError,
    canonical_diff,
    canonical_fingerprint,
    get_import_source,
)
from .lossiness import LossinessSeverity, LossItem
from .roundtrip_matrix import (
    MatchedDiff,
    MatrixCellStatus,
    classify_cell,
    import_adapter_for_emit,
    reconcile,
    reimport_emitted,
)

__all__ = [
    "ExportRoundtripRequest",
    "ExportRoundtripResponse",
    "run_export_roundtrip",
]

#: The apiome-rest package version stamped into reproduction provenance
#: (assignment style keeps the lowercase dunder import ruff-clean, matching
#: :mod:`app.export_projection`).
APIOME_VERSION = __version__

#: ``source_format`` label used when classifying the Studio's single-document cell.
#: The matrix labels cells by corpus source format; a Studio run has no corpus file,
#: so the label only appears inside the human-readable skip reason.
_STUDIO_SOURCE_LABEL = "studio"


class ExportRoundtripRequest(BaseModel):
    """A round-trip comparison request: source revision + chosen target + options."""

    model_config = ConfigDict(extra="forbid")

    artifact: str = Field(description="The artifact (project) id to round-trip.")
    version: Optional[str] = Field(
        default=None,
        description="Revision UUID, version label (``1.0.0``), or null for the latest revision.",
    )
    target: str = Field(
        description="Target emitter key (``openapi``) or format key (``openapi-3.1``).",
    )
    options: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Per-target emit options (MFX-1.4); null or empty applies the target defaults.",
    )
    min_severity: LossinessSeverity = Field(
        default=LossinessSeverity.INFO,
        description="Lowest loss severity that raises the advisory (MFX-2.4); does not affect "
        "the round-trip verdict, the report, or the reconciliation.",
    )


class ExportRoundtripResponse(BaseModel):
    """The round-trip comparison result the Export Studio renders (IXH-4.4).

    ``status`` speaks the IXH-1.7 matrix's vocabulary so a Studio run over a corpus
    entry reconciles with the published matrix cell:

    * ``pass`` — every empirical difference is explained by the fidelity report and
      nothing over-claims preservation;
    * ``fail`` — at least one unexplained difference or over-claim (a fidelity bug
      worth reporting), or the emitted artifact could not be re-imported at all;
    * ``unsupported`` — the comparison was **skipped with an explanation** (no import
      adapter can re-ingest the emit format, or the emitter is unavailable here).

    The three difference groups are exactly the reconciliation's: ``matched``
    (explained by the report — expected loss), ``unexplained`` (the report does not
    account for them), and ``overclaims`` (the report claimed preservation reality
    contradicts). The provenance fields (`emitter_version` / `apiome_version` /
    `registry_version` / fingerprints) give an unexplained difference's issue report
    its reproduction coordinates without shipping any source bytes.
    """

    model_config = ConfigDict(extra="forbid")

    artifact: str = Field(description="The artifact (project) id the round-trip ran for.")
    version: Optional[str] = Field(
        default=None, description="The version selector as requested (label, UUID, or null)."
    )
    version_record_id: str = Field(description="The resolved revision (``versions.id``).")
    version_label: Optional[str] = Field(
        default=None, description="The resolved revision's version label (e.g. ``1.0.0``)."
    )
    target: str = Field(description="The resolved target format key (e.g. ``openapi-3.1``).")
    emit_key: str = Field(description="The resolved emitter registry key (e.g. ``openapi``).")
    adapter_key: Optional[str] = Field(
        default=None,
        description="The import-adapter registry key that re-imported the emitted artifact; "
        "null when the comparison was skipped (``status: unsupported``).",
    )
    status: MatrixCellStatus = Field(
        description="The round-trip verdict (``pass`` / ``fail`` / ``unsupported``), in the "
        "IXH-1.7 matrix's vocabulary.",
    )
    reason: Optional[str] = Field(
        default=None,
        description="Human-readable explanation for a non-``pass`` status: why the comparison "
        "was skipped, what failed to re-import, or which differences are unaccounted for.",
    )
    diff_count: int = Field(
        default=0, description="Total empirical differences between source and re-import."
    )
    matched_count: int = Field(
        default=0, description="Differences the fidelity report explains (expected loss)."
    )
    matched: List[MatchedDiff] = Field(
        default_factory=list,
        description="Each explained difference paired with the fidelity finding that explains it.",
    )
    unexplained: List[CanonicalDiffEntry] = Field(
        default_factory=list,
        description="Differences no fidelity finding accounts for — a fidelity bug the user "
        "should report.",
    )
    overclaims: List[LossItem] = Field(
        default_factory=list,
        description="``OK`` findings whose construct empirically changed or vanished — the "
        "report over-claimed preservation.",
    )
    loss_drop: int = Field(default=0, description="``drop`` findings in the fidelity report.")
    loss_approx: int = Field(default=0, description="``approx`` findings in the fidelity report.")
    loss_synth: int = Field(default=0, description="``synth`` findings in the fidelity report.")
    loss_ok: int = Field(default=0, description="``ok`` findings in the fidelity report.")
    source_fingerprint: str = Field(
        description="Deterministic fingerprint of the source canonical model (reproduction "
        "coordinate for issue reports; carries no source content).",
    )
    reimported_fingerprint: Optional[str] = Field(
        default=None,
        description="Fingerprint of the re-imported canonical model; equals "
        "``source_fingerprint`` for a byte-honest round-trip. Null when the loop did not close.",
    )
    emitter_version: str = Field(description="The emitter implementation version.")
    apiome_version: str = Field(description="The apiome-rest package version that ran the loop.")
    registry_version: str = Field(description="The capability-registry snapshot version.")


def _target_entry(target_format: str, emit_key: str) -> Optional[EmitterTarget]:
    """Find the registry's :class:`EmitterTarget` for a resolved dispatch target.

    Args:
        target_format: The resolved emit format key (``openapi-3.1``).
        emit_key: The resolved emitter registry key (``openapi``).

    Returns:
        The matching registry entry, or ``None`` when the registry and the dispatch
        disagree (defensive; they are built from the same registrations).
    """
    for entry in describe_emit_targets():
        if entry.descriptor.format == target_format or entry.descriptor.key == emit_key:
            return entry
    return None


def run_export_roundtrip(
    source: ExportSource,
    target: str,
    *,
    version: Optional[str] = None,
    options: Optional[Dict[str, Any]] = None,
    min_severity: LossinessSeverity = LossinessSeverity.INFO,
) -> ExportRoundtripResponse:
    """Emit ``source`` to ``target``, re-import, diff against the source, reconcile.

    One explicit, bounded run: a single read-only emit (no artifact, job, or
    field-identity rows persisted), one re-import through the matching adapter, one
    :func:`~app.import_source.canonical_diff`, and one
    :func:`~app.roundtrip_matrix.reconcile` against the **same** fidelity report the
    dispatch computed for this emit — so the verdict and the Studio's fidelity
    surfaces describe one snapshot.

    Args:
        source: The loaded export source (canonical model + resolved coordinates).
        target: Target emitter ``key`` (``openapi``) or format key (``openapi-3.1``).
        version: The caller's version selector, echoed back verbatim.
        options: Per-target emit options; ``None``/empty applies the target defaults.
        min_severity: Advisory threshold passed through to the dispatch; does not
            affect the verdict.

    Returns:
        The full :class:`ExportRoundtripResponse` (``pass`` / ``fail`` /
        ``unsupported``, the three difference groups, and reproduction provenance).

    Raises:
        ExportError: When ``target`` does not resolve, its options are invalid, or
            the emitter produced no document — the same taxonomy the dispatch and
            verify surfaces raise, mapped to HTTP by the route.
    """
    # confirm=True: a severe conversion is *measured* here, not blocked — the round-trip
    # exists precisely to show what a lossy conversion does. persistence=None keeps the
    # emit read-only. (Same stance as the verify surface, MFX-42.5.)
    dispatch = dispatch_from_source(
        source,
        target,
        options=options,
        min_severity=min_severity,
        dry_run=False,
        confirm=True,
        persistence=None,
    )
    descriptor = dispatch.fidelity.target
    emitter_cls = get_emitter(descriptor.format)
    emitter_version = getattr(emitter_cls, "version", "unknown") if emitter_cls else "unknown"
    source_fingerprint = canonical_fingerprint(source.api)
    report = dispatch.fidelity.report

    base: Dict[str, Any] = {
        "artifact": dispatch.artifact,
        "version": version,
        "version_record_id": dispatch.version_record_id,
        "version_label": dispatch.version_label,
        "target": dispatch.target,
        "emit_key": descriptor.key,
        "loss_drop": report.kind_counts.get("drop", 0),
        "loss_approx": report.kind_counts.get("approx", 0),
        "loss_synth": report.kind_counts.get("synth", 0),
        "loss_ok": report.kind_counts.get("ok", 0),
        "source_fingerprint": source_fingerprint,
        "emitter_version": emitter_version,
        "apiome_version": APIOME_VERSION,
        "registry_version": REGISTRY_VERSION,
    }

    # Gate: can the loop close at all? classify_cell is the matrix's own gate, so the
    # skip reason here is word-for-word the reason the published 1.7 grid records.
    entry = _target_entry(dispatch.target, descriptor.key)
    if entry is not None:
        classified = classify_cell(_STUDIO_SOURCE_LABEL, entry)
        if classified is not None:
            return ExportRoundtripResponse(
                **base,
                status=MatrixCellStatus.UNSUPPORTED,
                reason=classified.reason,
            )
    adapter_key = import_adapter_for_emit(descriptor.key, descriptor.format)
    if adapter_key is None:
        # Reachable only if the registry entry vanished between dispatch and gate.
        return ExportRoundtripResponse(
            **base,
            status=MatrixCellStatus.UNSUPPORTED,
            reason=(
                f"No import adapter can re-import emit format {descriptor.format!r} "
                f"(emit key {descriptor.key!r})."
            ),
        )
    adapter = get_import_source(adapter_key)
    if adapter is None:
        return ExportRoundtripResponse(
            **base,
            status=MatrixCellStatus.UNSUPPORTED,
            reason=f"Import adapter {adapter_key!r} unresolved in this runtime.",
        )

    # A real (non-dry-run) dispatch that returned without raising always emits.
    assert dispatch.emit is not None

    try:
        reimported = reimport_emitted(
            adapter,
            dispatch.emit,
            multi_file=descriptor.multi_file,
            source_label=f"roundtrip:{dispatch.artifact}->{descriptor.key}",
        )
    except (ImportSourceError, ValueError, TypeError) as exc:
        # The emitted artifact could not be re-ingested by its own matching adapter —
        # itself a fidelity bug, reported as a failure rather than raised, so the
        # Studio can render it (and its issue-report path) like any other failure.
        return ExportRoundtripResponse(
            **base,
            adapter_key=adapter_key,
            status=MatrixCellStatus.FAIL,
            reason=f"Re-import failed: {exc}",
        )

    diff = canonical_diff(source.api, reimported)
    reconciliation = reconcile(diff, report)

    if reconciliation.ok:
        status = MatrixCellStatus.PASS
        reason = None
    else:
        status = MatrixCellStatus.FAIL
        parts: List[str] = []
        if reconciliation.unexplained:
            keys = ", ".join(
                f"{e.change.value} {e.entity} {e.key!r}" for e in reconciliation.unexplained
            )
            parts.append(f"unexplained: {keys}")
        if reconciliation.overclaims:
            keys = ", ".join(f"OK {item.construct_key!r}" for item in reconciliation.overclaims)
            parts.append(f"overclaim: {keys}")
        reason = "; ".join(parts)

    return ExportRoundtripResponse(
        **base,
        adapter_key=adapter_key,
        status=status,
        reason=reason,
        diff_count=len(diff.entries),
        matched_count=len(reconciliation.matched),
        matched=list(reconciliation.matched),
        unexplained=list(reconciliation.unexplained),
        overclaims=list(reconciliation.overclaims),
        reimported_fingerprint=canonical_fingerprint(reimported),
    )

File: apiome-ui/src/app/components/ade/dashboard/export/exportRoundtrip.ts
/**
 * Round-trip comparison envelope + presentation + issue-report helpers (IXH-4.4, #5112).
 *
 * The Export Studio's round-trip action answers "is this export honest?" empirically: the
 * server emits the source revision, re-imports the artifact through the matching import
 * adapter, diffs the re-imported canonical model against the source, and reconciles every
 * difference against the fidelity report (`POST /api/export/roundtrip`, the same loop the
 * IXH-1.7 conformance matrix runs in CI). This module mirrors that response field-for-field
 * so it deserialises directly, and adds the pure logic the panel needs:
 *
 *  - verdict presentation (`pass` / `fail` / `unsupported`) stated in words + glyph before
 *    colour (the Studio's house rule: meaning is never encoded by colour alone);
 *  - labels for the difference groups — differences the fidelity report explains are
 *    **expected**; unexplained differences and over-claims flag a fidelity bug;
 *  - the one-click issue-report builder for unexplained differences: a prefilled GitHub
 *    new-issue URL carrying the reproduction coordinates (source/target/options with
 *    credential-shaped option keys stripped, fingerprints, component versions) and **no
 *    source bytes**.
 *
 * Everything here is pure (no React, no fetch) so it unit-tests directly — mirroring
 * `./exportVerify.ts` and `./exportFidelityPreview.ts`.
 */

import type { StatusTone } from '@/app/components/ui/statusVocabulary';
import type { LossItem } from './exportFidelityPreview';
import { stripSecretOptions } from './exportStudioUrlState';

/**
 * The round-trip verdict, in the IXH-1.7 matrix's vocabulary (mirrors Python
 * `MatrixCellStatus`; the on-demand endpoint returns only the first three).
 */
export type RoundtripStatus = 'pass' | 'fail' | 'unsupported' | 'skipped' | 'xfail';

/** How one canonical entity differed after the round-trip (mirrors Python `DiffChangeKind`). */
export type RoundtripChangeKind = 'added' | 'removed' | 'changed';

/** One empirical difference between the source and the re-imported model. */
export interface RoundtripDiffEntry {
  /** The entity granularity: `root` / `service` / `operation` / `type` / `channel`. */
  entity: string;
  /** The entity's stable canonical key (empty string for the artifact root). */
  key: string;
  /** How it differed: added / removed / changed. */
  change: RoundtripChangeKind;
}

/** One explained difference: the diff entry paired with the fidelity finding covering it. */
export interface RoundtripMatchedDiff {
  /** The empirical difference. */
  entry: RoundtripDiffEntry;
  /** The fidelity finding that explains it (same shape as the fidelity report's items). */
  finding: LossItem;
}

/** The `POST /api/export/roundtrip` response (mirrors REST `ExportRoundtripResponse`). */
export interface ExportRoundtripResponse {
  /** The artifact (project) id the round-trip ran for. */
  artifact: string;
  /** The version selector as requested (label, UUID, or null for latest). */
  version?: string | null;
  /** The resolved revision record id. */
  version_record_id: string;
  /** The resolved revision's version label, e.g. `"1.2.0"`. */
  version_label?: string | null;
  /** The resolved target format key (e.g. `openapi-3.1`). */
  target: string;
  /** The resolved emitter registry key (e.g. `openapi`). */
  emit_key: string;
  /** The import adapter that re-imported the artifact; null when the comparison was skipped. */
  adapter_key?: string | null;
  /** The verdict: pass / fail / unsupported. */
  status: RoundtripStatus;
  /** Why a non-pass verdict happened (skip explanation, re-import failure, unexplained keys). */
  reason?: string | null;
  /** Total empirical differences between the source and the re-import. */
  diff_count: number;
  /** Differences the fidelity report explains (expected loss). */
  matched_count: number;
  /** Each explained difference paired with its finding. */
  matched: RoundtripMatchedDiff[];
  /** Differences no fidelity finding accounts for — a fidelity bug worth reporting. */
  unexplained: RoundtripDiffEntry[];
  /** `ok` findings whose construct empirically changed or vanished (over-claimed preservation). */
  overclaims: LossItem[];
  /** `drop` findings in the fidelity report. */
  loss_drop: number;
  /** `approx` findings in the fidelity report. */
  loss_approx: number;
  /** `synth` findings in the fidelity report. */
  loss_synth: number;
  /** `ok` findings in the fidelity report. */
  loss_ok: number;
  /** Deterministic fingerprint of the source canonical model (no source content). */
  source_fingerprint: string;
  /** Fingerprint of the re-imported model; equals the source's for a byte-honest loop. */
  reimported_fingerprint?: string | null;
  /** The emitter implementation version. */
  emitter_version: string;
  /** The apiome-rest package version that ran the loop. */
  apiome_version: string;
  /** The capability-registry snapshot version. */
  registry_version: string;
}

/** How a verdict is presented: the meaning in words and a glyph before any colour. */
export interface RoundtripStatusPresentation {
  /** Short verdict label (e.g. `Round trip verified`). */
  label: string;
  /** A one-line sentence stating what the verdict means for this artifact. */
  sentence: string;
  /** A text glyph reinforcing the verdict (never the only signal). */
  glyph: string;
  /**
   * The verdict banner's tone (colour arrives last, after words + glyph).
   *
   * HIVE-8.3 (#5329): a tone name rather than the four Tailwind palette triples this used to
   * carry, so the banner follows all nine themes through `ui/Alert`.
   */
  tone: StatusTone;
}

/**
 * Present a round-trip verdict (words + glyph first, colour last — the house rule).
 *
 * @param response The settled round-trip response.
 * @returns The banner label, sentence, glyph, and classes for the verdict.
 */
export function roundtripStatusPresentation(
  response: ExportRoundtripResponse,
): RoundtripStatusPresentation {
  switch (response.status) {
    case 'pass':
      return response.diff_count === 0
        ? {
            label: 'Round trip verified',
            sentence:
              'The emitted artifact re-imported to an identical canonical model — no differences at all.',
            glyph: '✓',
            tone: 'ok',
          }
        : {
            label: 'Round trip verified',
            sentence:
              'Every difference between the source and the re-imported artifact is explained by the fidelity report.',
            glyph: '✓',
            tone: 'ok',
          };
    case 'unsupported':
      return {
        label: 'Comparison skipped',
        sentence:
          response.reason ??
          'No import adapter can re-import this format, so the round trip cannot be closed.',
        glyph: '−',
        tone: 'neutral',
      };
    default:
      return {
        label: 'Fidelity report incomplete',
        sentence:
          'The round trip found differences the fidelity report does not account for — likely a fidelity bug worth reporting.',
        glyph: '✗',
        tone: 'danger',
      };
  }
}

/** Human label for a diff change kind (e.g. `Removed`). */
export function changeKindLabel(change: RoundtripChangeKind): string {
  switch (change) {
    case 'added':
      return 'Added';
    case 'removed':
      return 'Removed';
    default:
      return 'Changed';
  }
}

/**
 * The tone per diff change kind (colour after the label, never instead of it).
 *
 * HIVE-8.3 (#5329) replaced the three Tailwind palette pairs this returned with vocabulary
 * tones, so a "Removed" here is the same rose as a dropped construct in the manifest tree.
 */
export function changeKindTone(change: RoundtripChangeKind): StatusTone {
  switch (change) {
    case 'added':
      return 'violet';
    case 'removed':
      return 'rose';
    default:
      return 'accent';
  }
}

/** Render one diff entry as `entity key` prose (the root entity has an empty key). */
export function diffEntryLabel(entry: RoundtripDiffEntry): string {
  return entry.key === '' ? 'artifact root' : `${entry.entity} ${entry.key}`;
}

/**
 * One line summarising a settled round trip for the panel's status region — e.g.
 * `3 differences · 3 explained by the fidelity report` or
 * `4 differences · 2 unexplained · 1 over-claim`.
 */
export function summarizeRoundtrip(response: ExportRoundtripResponse): string {
  if (response.status === 'unsupported') return 'Comparison skipped.';
  if (response.diff_count === 0 && response.overclaims.length === 0) {
    return 'No differences — the re-imported model is identical to the source.';
  }
  const parts = [`${response.diff_count} difference${response.diff_count === 1 ? '' : 's'}`];
  if (response.matched_count > 0) {
    parts.push(`${response.matched_count} explained by the fidelity report`);
  }
  if (response.unexplained.length > 0) {
    parts.push(`${response.unexplained.length} unexplained`);
  }
  if (response.overclaims.length > 0) {
    parts.push(
      `${response.overclaims.length} over-claim${response.overclaims.length === 1 ? '' : 's'}`,
    );
  }
  return parts.join(' · ');
}

/** Where round-trip fidelity bugs are filed. */
export const ROUNDTRIP_ISSUE_BASE = 'https://github.com/apiome/apiome/issues/new';

/**
 * GitHub caps a `/issues/new` URL well below 8 KiB; beyond it the prefill is dropped
 * entirely. The body is clipped (with a note) before encoding so the link always works.
 */
export const ISSUE_BODY_MAX_LENGTH = 5_000;

/** The inputs the issue report is built from. */
export interface RoundtripIssueInput {
  /** The settled (failing) round-trip response. */
  response: ExportRoundtripResponse;
  /** The chosen target's human label (e.g. `OpenAPI 3.1`). */
  targetLabel: string;
  /** The non-default option overrides the run was configured with, or null. */
  options?: Record<string, unknown> | null;
}

/** A prefilled issue report: the title/body and the URL that carries them. */
export interface RoundtripIssueReport {
  /** The issue title. */
  title: string;
  /** The issue body (Markdown), before URL encoding. */
  body: string;
  /** The prefilled GitHub new-issue URL. */
  url: string;
  /** Option keys withheld from the report because they look like credentials. */
  redactedOptionKeys: string[];
}

/**
 * Build the one-click issue report for a failed round trip (IXH-4.4 acceptance: unexplained
 * differences offer an issue-report path carrying the reproduction inputs).
 *
 * The reproduction block is coordinates, not content: artifact/revision ids, target and
 * adapter keys, the non-default options (credential-shaped keys stripped and named as
 * withheld), the model fingerprints, and the component versions. No source bytes, option
 * secrets, or emitted content ever ride in the URL.
 *
 * @param input The failing response + the target's label + the configured options.
 * @returns The prefilled report; `url` opens GitHub's new-issue form with it.
 */
export function buildRoundtripIssueReport(input: RoundtripIssueInput): RoundtripIssueReport {
  const { response, targetLabel } = input;
  const { safe, redacted } = stripSecretOptions(input.options ?? null);
  const title = `Round-trip fidelity gap: ${response.emit_key} leaves ${
    response.unexplained.length > 0
      ? `${response.unexplained.length} difference${response.unexplained.length === 1 ? '' : 's'} unexplained`
      : `${response.overclaims.length} over-claim${response.overclaims.length === 1 ? '' : 's'}`
  }`;

  const lines: string[] = [
    '### What happened',
    '',
    `An on-demand round-trip comparison (Export Studio, IXH-4.4) of an export to **${targetLabel}** ` +
      'found differences the fidelity report does not account for.',
    '',
    '### Unexplained differences',
    '',
  ];
  if (response.unexplained.length > 0) {
    for (const entry of response.unexplained) {
      lines.push(`- ${changeKindLabel(entry.change)} ${diffEntryLabel(entry)}`);
    }
  } else {
    lines.push('- (none — see over-claims)');
  }
  if (response.overclaims.length > 0) {
    lines.push('', '### Over-claimed preservation', '');
    for (const item of response.overclaims) {
      lines.push(`- \`${item.construct}\` was reported preserved (ok) but changed or vanished`);
    }
  }
  if (response.reason) {
    lines.push('', '### Reported reason', '', '```', response.reason, '```');
  }
  lines.push(
    '',
    '### Reproduction coordinates',
    '',
    `- Target: \`${response.target}\` (emit key \`${response.emit_key}\`)`,
    `- Re-import adapter: \`${response.adapter_key ?? 'n/a'}\``,
    `- Artifact id: \`${response.artifact}\``,
    `- Revision: \`${response.version_record_id}\`${
      response.version_label ? ` (version ${response.version_label})` : ''
    }`,
    `- Options: ${safe ? `\`${JSON.stringify(safe)}\`` : 'target defaults'}${
      redacted.length > 0 ? ` (withheld credential-shaped keys: ${redacted.join(', ')})` : ''
    }`,
    `- Source fingerprint: \`${response.source_fingerprint}\``,
    `- Re-imported fingerprint: \`${response.reimported_fingerprint ?? 'n/a'}\``,
    `- Versions: apiome-rest \`${response.apiome_version}\`, emitter \`${response.emitter_version}\`, registry \`${response.registry_version}\``,
    '',
    '_Reported from the Export Studio round-trip comparison. No source content is included._',
  );

  let body = lines.join('\n');
  if (body.length > ISSUE_BODY_MAX_LENGTH) {
    body = `${body.slice(0, ISSUE_BODY_MAX_LENGTH)}\n\n_(clipped — full detail available by re-running the round trip)_`;
  }
  const url = `${ROUNDTRIP_ISSUE_BASE}?title=${encodeURIComponent(title)}&labels=${encodeURIComponent(
    'export,diff',
  )}&body=${encodeURIComponent(body)}`;
  return { title, body, url, redactedOptionKeys: redacted };
}
